sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -63,12 +73,342 @@ abstract class GpuShuffleBlockResolverBase(
+}
+
+abstract class RapidsShuffleThreadedWriterBase[K, V](
+    blockManager: BlockManager,
+    handle: BypassMergeSortShuffleHandle[K, V],
+    mapId: Long,
+    sparkConf: SparkConf,
+    writeMetrics: ShuffleWriteMetricsReporter,
+    shuffleExecutorComponents: ShuffleExecutorComponents)
+      extends ShuffleWriter[K, V]
+        with RapidsShuffleWriterShimHelper
+        with Arm
+        with Logging {
+
+  private var myMapStatus: Option[MapStatus] = None
+
+  private val dep: ShuffleDependency[K, V, V] = handle.dependency
+  private val shuffleId = dep.shuffleId
+  private val partitioner = dep.partitioner
+  private val numPartitions = partitioner.numPartitions
+  private val serializer = dep.serializer.newInstance()
+  private val transferToEnabled = sparkConf.getBoolean("spark.file.transferTo", true)
+  private val fileBufferSize = sparkConf.get(config.SHUFFLE_FILE_BUFFER_SIZE).toInt * 1024
+  /**
+   * Are we in the process of stopping? Because map tasks can call stop() with success = true
+   * and then call stop() with success = false if they get an exception, we want to make sure
+   * we don't try deleting files, etc twice.
+   */
+  private var stopping = false
+
+  val diskBlockObjectWriters = new mutable.HashMap[Int, (Int, DiskBlockObjectWriter)]()
+
+  override def write(records: Iterator[Product2[K, V]]): Unit = {
+    withResource(new NvtxRange("ThreadedWriter.write", NvtxColor.RED)) { _ =>
+      withResource(new NvtxRange("compute", NvtxColor.GREEN)) { _ =>
+        var errorOcurred: Throwable = null
+        val mapOutputWriter = shuffleExecutorComponents.createMapOutputWriter(
+          shuffleId,
+          mapId,
+          numPartitions)
+        try {
+          val partLengths = if (!records.hasNext) {
+            commitAllPartitions(mapOutputWriter, true /*empty checksum*/)
+          } else {
+            // per reduce partition id
+            // open all the writers ahead of time (Spark does this already)
+            val openStartTime = System.nanoTime()
+            (0 until numPartitions).map { i =>
+              val (blockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
+              val writer: DiskBlockObjectWriter = blockManager.getDiskWriter(
+                blockId, file, serializer, fileBufferSize, writeMetrics)
+              setChecksumIfNeeded(writer, i) // spark3.2.0+
+
+              // Places writer objects at round robin slot numbers apriori
+              // this choice is for simplicity but likely needs to change so that
+              // we can handle skew better
+              val slotNum = RapidsShuffleInternalManagerBase.getNextSlot
+              diskBlockObjectWriters.put(i, (slotNum, writer))
+            }
+            writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
+
+            // we call write on every writer for every record in parallel
+            val writeFutures = new ArrayBuffer[Future[Unit]]
```

Comment:
Should this be a queue-like structure since we're going to be popping the head element all the time?  ArrayBuffer isn't great for that, copying all the values down for each removal.

Reply:
switched to mutable.Queue